src/sync/impl/sync_file.cpp

```diff
@@ -0,0 +1,316 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "sync/impl/sync_file.hpp"
+
+#include <realm/util/file.hpp>
+
+using File = realm::util::File;
+
+namespace realm {
+
+namespace {
+
+uint8_t value_of_hex_digit(char hex_digit)
+{
+    if (hex_digit >= '0' && hex_digit <= '9') {
+        return hex_digit - '0';
+    } else if (hex_digit >= 'A' && hex_digit <= 'F') {
+        return 10 + hex_digit - 'A';
+    } else if (hex_digit >= 'a' && hex_digit <= 'f') {
+        return 10 + hex_digit - 'a';
+    } else {
+        throw std::invalid_argument("Cannot get the value of a character that isn't a hex digit.");
+    }
+}
+
+bool filename_is_reserved(const std::string& filename) {
+    return (filename == "." || filename == "..");
+}
+
+bool character_is_unreserved(char character)
+{
+    bool is_capital_letter = (character >= 'A' && character <= 'Z');
+    bool is_lowercase_letter = (character >= 'a' && character <= 'z');
+    bool is_number = (character >= '0' && character <= '9');
+    bool is_allowed_symbol = (character == '-' || character == '_' || character == '.');
+    return is_capital_letter || is_lowercase_letter || is_number || is_allowed_symbol;
+}
+
+char decoded_char_for(const std::string& percent_encoding, size_t index)
+{
+    if (index+2 >= percent_encoding.length()) {
+        throw std::invalid_argument("Malformed string: not enough characters after '%' before end of string.");
+    }
+    REALM_ASSERT(percent_encoding[index] == '%');
+    return (16*value_of_hex_digit(percent_encoding[index + 1])) + value_of_hex_digit(percent_encoding[index + 2]);
+}
+
+} // (anonymous namespace)
+
+namespace util {
+
+void remove_nonempty_dir(const std::string& path)
+{
+#ifdef _WIN32
+    static_assert(false, "Not implemented for Win32 yet");
+#else
+    // Open the directory and list all the files.
+    DIR *dir_listing = opendir(path.c_str());
+    if (!dir_listing) {
+        return;
+    }
+    while (struct dirent *file = readdir(dir_listing)) {
```

Comment:
`auto cleanup = util::make_scope_exit([=]() noexcept { closedir(dir_listing); });` here rather than the call to `closedir()` below will make it so that this doesn't leak the directory handle if something below throws.